### DISCUSSION TITLE
Can you use flow for advanced validations

### DISCUSSION THREAD

**Discussion by sptv2reg** [directus/directus#16802] [Archive]
Hi

Let me start with the usual stuff - what a great product you've made here, thanks.

I've got a pretty advanced validation of a string field. The validation requires me to make a http request to an external service to check certain things. I can do that with the http-module.

However, is it possible to trigger this on insert/update and return an error message that result in the same red alertbox at the top as the native validation rules?

**Comment by licitdev** [comment]
Here's an example of how you can tap on the existing validation errors.

https://user-images.githubusercontent.com/26413686/207580853-60f9269b-7b83-4a85-be11-e6dae59a6e25.mov

```js
module.exports = function registerHooks({ filter }, { exceptions }) {
	filter('test.items.create', async (input) => {
		if (!input.name) {
			throw new exceptions.FailedValidationException({
				message: 'Missing name',
				path: ['name'],
				type: 'custom.name.invalid',
				context: {
					invalids: [''],
				},
			});
		} else if (input.name !== 'test') {
			throw new exceptions.FailedValidationException({
				message: 'Invalid name',
				path: ['name'],
				type: 'custom.name.invalid',
				context: {
					invalids: [input.name],
				},
			});
		}
	});
};
```

As the naming of `type` and values required in `context`, please refer to https://github.com/directus/directus/blob/c98e99be74fad8c33382c1614b4a776817150ef7/packages/shared/src/exceptions/failed-validation.ts for more information.

**Reply by sptv2reg** [reply]
Thank you. This is an extension right? Is it possible to execute this in the context of a directus Flow?

**Reply by licitdev** [reply]
Yes, the above is an extension. In flows, you do not have access to the exception, but you may throw the error values directly as follows. The below example is to be used in a `Run Script` operation.

```js
module.exports = async function (data) {
	const input = data.$trigger.payload;

	if (!input.name) {
		throw {
			message: 'Invalid name',
			extensions: {
				code: 'FAILED_VALIDATION',
				field: 'name',
				type: 'nempty',
			},
		};
	} else if (input.name !== 'test') {
		throw {
			message: 'Invalid name',
			extensions: {
				code: 'FAILED_VALIDATION',
				field: 'name',
				type: 'neq',
				invalid: input.name,
			},
		};
	}
};
```

**Reply by sptv2reg** [reply]
This works. Excellent, thanks!

**Reply by sptv2reg** [reply]
A short followup - It doesn't seem that "message" is doing anything. If you use "neq" it will output "Name: Value can't be" no matter what I write in message.

**Reply by licitdev** [reply]
The validation error reads into the `type` for the error message. Custom validation error messages isn't currently supported.

**Reply by maisur** [reply]
If you set the `code` using http code. The error will show up in modal and using the custom message.
```javascript
throw {
	message: 'Invalid image',
	extensions: {
		code: 400,
		field: 'name',
		type: 'neq',
		invalid: input.name,
	},
};
```

![image](https://github.com/directus/directus/assets/47018830/7c5d106d-85e1-472b-a3da-b0462fbbe943)
